### ISSUE TITLE
Possible memory leak in Java driver when reusing a Connection

### ISSUE THREAD

**Issue by azabujuban** [rethinkdb/rethinkdb#5980]
When only one connection is kept open and reused multiple times the `cursorCache` map is growing and the `Cursor` objects are never removed from it.

**Comment by bchavez** [comment]
**C# driver**

https://github.com/bchavez/RethinkDb.Driver/blob/master/Source/RethinkDb.Driver/Net/Cursor.cs#L251
```
void Shutdown(string reason)
.... conn.RemoveFromCache(this.Token);
```

**Java driver**, no cache removal on that critical piece of code:

https://github.com/rethinkdb/rethinkdb/blob/next/drivers/java/src/main/java/com/rethinkdb/net/Cursor.java#L49

I ***maybe*** be related,  [cursor.close()](https://github.com/rethinkdb/rethinkdb/blob/next/drivers/java/src/main/java/com/rethinkdb/net/Cursor.java#L49) -> [conn. stop()](https://github.com/rethinkdb/rethinkdb/blob/next/drivers/java/src/main/java/com/rethinkdb/net/Cursor.java#L54) -> [conn.runQueryNoreply()](https://github.com/rethinkdb/rethinkdb/blob/next/drivers/java/src/main/java/com/rethinkdb/net/Connection.java#L327) and https://github.com/rethinkdb/rethinkdb/issues/5938

PR on it's way for this obvious one, but maybe not #5938 since it's a bit more work to sync-up.

:collision: :dizzy: [***"Crashing, hit a wall. Right now I need a miracle..."***](https://www.youtube.com/watch?v=s8XIgR5OGJc)

**Comment by danielmewes** [comment]
@azabujuban Thanks for reporting this and tracking down the issue in the code.

@bchavez You're the best! :) Thanks for the fast patch.

**Comment by deontologician** [comment]
:+1: lgtm, merging

**Comment by danielmewes** [comment]
We just released a new Java driver version 2.3.3 with this fix.
